precompiles/randomness/src/lib.rs

```diff
@@ -65,37 +76,42 @@ pub fn log_fulfillment_failed(address: impl Into<H160>) -> Log {
 
 /// Reverts if fees and gas_limit are not sufficient to make subcall and cleanup
 fn ensure_can_provide_randomness<Runtime>(
-	code_address: H160,
-	gas_limit: u64,
+	remaining_gas: u64,
+	request_gas_limit: u64,
 	request_fee: BalanceOf<Runtime>,
-	clean_up_cost: u64,
+	subcall_overhead_gas_costs: u64,
+	prepare_and_finish_fulfillment_gas_cost: u64,
 ) -> EvmResult<()>
 where
 	Runtime: pallet_randomness::Config + pallet_evm::Config,
 	BalanceOf<Runtime>: Into<U256>,
 {
-	// assert fee > gasLimit * base_fee
-	let gas_limit_as_u256: U256 = gas_limit.into();
+	let request_gas_limit_with_overhead = request_gas_limit
+		.checked_add(subcall_overhead_gas_costs)
+		.ok_or(revert(
+			"overflow when computing request gas limit + overhead",
+		))?;
+
+	// Ensure precompile have enough gas to perform subcall with the overhead.
+	if remaining_gas < request_gas_limit_with_overhead {
+		return Err(revert("not enough gas to perform the call"));
+	}
+
+	// Ensure request fee can pay for this.
+	let total_refunded_gas = prepare_and_finish_fulfillment_gas_cost
+		.checked_add(request_gas_limit_with_overhead)
+		.ok_or(revert("overflow when computed max amount of refunded gas"))?;
+
+	let total_refunded_gas: U256 = total_refunded_gas.into();
 	let (base_fee, _) = <Runtime as pallet_evm::Config>::FeeCalculator::min_gas_price();
```

Comment:
in theory we should include the second parameter (it is the amount of weight for calling `min_gas_price()` itself). we currently use a value of 0 for this, though.